### ISSUE TITLE
Add automatic list key generation (autokey)

### ISSUE THREAD

**Issue by abierman** [netmod-wg/yang-next#141]
There are many use-cases where the list key does not contain any semantics other than being a unique identifier. 

- In this mode, the designer does not want to define or implement the list key management.
- The client does not want to provide a new unique list key by first locking the config and retrieving all existing entries to find an unused entry
- The current protocol operations do not support this feature and need improvements
- The key is a uint32 number from 1 .. max
- The value 0 is reserved for indicate no key
- Not allowed for ordered-by user, or no insert operation allowed for create
- The autokey is un-named so it does not conflict with any siblings, or could use reserved name like "\_id\_"
- This is not the same as position-based (no key-stmt at all) because autokey does not renumber for any reason
- The server assigns an unused value for the ID and the protocol operation enhancements allow this operation and convey the assigned ID to the client
```
list shopping-list {
    autokey;
    // list child nodes
}
```

**Comment by BalazsLengyel** [comment]
Importance high. 3GPP would be happy to see this. 
The key should still be readable, visible.
This seems to be dependent on the protocol too.

**Comment by janlindblad** [comment]
I am strongly opposed to the servers inventing their own configuration. Others management protocols support this idea, and the implications are far reaching, complex and inefficient. I can provide examples of what this kind of world looks like, but it's a little to big for a comment here.

**Comment by BalazsLengyel** [comment]
While I see the dangers, this is constantly raised as an advantage of REST, that the client doesn't need to select a free key value.

**Comment by plajjan** [comment]
I completely agree with Jan that the implications are very far reaching and I too am strongly opposed to this. It would completely change the way NETCONF / YANG works, where we say things like the server should not mutate the configuration, only the client is allowed to. Auto-key would change this upside down and now you could no longer know what config to remove that you once wrote. I think this property of knowing what you wrote so that you can remove the same thing is crucial and something that sets NETCONF / YANG apart from other architectures and it enables us to write truly declarative configuration systems.

With cloud orchestration tools, I believe it's a common practice for the client to create a resource and then grab the ID that was allocated, which it then needs to keep as part of local state. If that state is ever lost, we also loose track of our resources. We need those IDs so that when we modify or remove something, we know what resources in the cloud to clean up or modify. With NETCONF, I don't need that. I can write pure functions and given the same input, I can compute the same configuration, and thus a delta on the input to my automation system can cleanly compute a diff on the output side. Much more elegant!

**Comment by abierman** [comment]
As stated in the original post, there needs to be a protocol operation change
to allow this to be done in a way that returned the created IDs to the client.
Nobody is suggesting the current 'edit-config' which returns 'ok' or 'rpc-error' can support autokey.

This sort of thing is already being done with RPC operations and template expansions.
The IDs are not lost. They are visible in the server if the list is retrieved.

There are already proprietary workarounds so importance is low I guess

**Comment by BalazsLengyel** [comment]
"where we say things like the server should not mutate the configuration"
That is not prescribed in YANG. (show me it in RFC 7950) The server is allowed to change the configuration.

**Comment by abierman** [comment]
I do not get this work on "system" config, and the effort to keep <running> "pure" and free of "system edits".

There are many ways to convey config changes to the server.  edit-config, YANG Patch, and many others,
including proprietary mechanisms like templates and automation rules. There is no requirement that every
edit mechanism work the same or be designed with the same features.

I wonder if autokey is just a protocol detail, and not a YANG property.
To be a YANG property it must apply to every instance of every implementation.
There are many YANG lists where the key is an arbitrary number or string.
This does not really change for autokey.

**Comment by kwatsen** [comment]
Jan: strongly oppose
Kent: also unsure if this is a good idea
Jan: if entirely optional to clients, maybe okay, but experience is that its hard
Andy: this may not be a YANG issue (could be only a protocol effort)
Jan: is creating 100 interfaces, but 37 has a problem, how does a server report which had the problem
Andy: could report by position
Kent: position only valid for "order-by user"

**Comment by kwatsen** [comment]
Rob: this needs more discussion!

**Comment by plajjan** [comment]
@BalazsLengyel I believe idiomatic NETCONF / YANG clients should not mutate their own configuration. NETCONF is specified both as a forwarding looking technology but also one to envelop existing technology, which is why we had CLI wrapped in NETCONF in the early days. Clearly we could not forbid what devices were already doing. I think it's telling that in NMDA, we specifyy that system learned config does not go into the running datastore but into the operational one. This leaves running to be controlled by clients.

Going back to the roots of NETCONF & YANG; declarative, composable, reproducible & verifiable. That is natural when the server does not mutate its own config and on the contrary, quite hard to achieve if we allow the server to mutate.

autokeys requires that the systems start maintaining state in order to persist the retrieved IDs. Currently in NETCONF, I can easily invert the creation of a list element since I always know the key. The current approach is declarative, it is reproducible and make it easy to verify configuration consistency. Adding autokeys makes all of that much harder.

YANG list with arbitrary numbers or strings as keys are still known to the client. This is the key difference. This isn't about semantic meaning of the key, it is simply knowing the key so that the client can use it.

If we write some configuration to a device but it is modified into something else, how can we attribute that configuration? If we are to audit our network, how can we verify anything between our templates of expected configuration and the actual configuration of the devices when it by design is modified along the way?
